### DISCUSSION TITLE
How can I investigate hanging `composer dump-autoload` for 30 minutes?

### DISCUSSION THREAD

**Discussion by fluffycondor** [composer/composer#11487] [Q&A / Support] [answered]
Hi,

We have some legacy project that uses a classmap for autoloading because namespaces are a real mess there. There are 3 folders to scan and 4100 .php files inside in total.
When I run `composer dump-autoload` on our development server it runs successfully just in 2 seconds.
When our devops execute it in production, it runs for 30 minutes on the same codebase, the same git branch:
```
$ composer dump-autoload --profile -vvvvv
Running 2.5.5 (2023-03-21 11:50:05) with PHP 8.0.15 on Linux / 5.4.17-2136.316.7.el8uek.x86_64
[2.1MiB/0.01s] Reading ./composer.json (/var/www/project/composer.json)
[2.2MiB/0.02s] Loading config file ./composer.json (/var/www/project/composer.json)
[2.4MiB/0.02s] Checked CA file /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem: valid
[2.3MiB/0.04s] Executing command (/var/www/project): 'git' 'branch' '-a' '--no-color' '--no-abbrev' '-v'
[2.8MiB/0.11s] Failed to initialize global composer: Composer could not find the config file: /user/.config/composer/composer.json

[2.8MiB/0.11s] Reading ./composer.lock (/var/www/project/composer.lock)
[3.4MiB/0.11s] Reading /var/www/project/vendor/composer/installed.json
[4.3MiB/0.14s] Generating autoload files

[6.5MiB/1698.84s] Warning: Ambiguous class resolution, "Foo" was found in both "..." and "...", the first will be used.
.... 10 more same warnings ....
[5.0MiB/1698.90s] Generated autoload files
[4.4MiB/1698.90s] Memory usage: 4.43MiB (peak: 8.45MiB), time: 1698.9s
```
Also Internet connection is restricted on our production server, but as far as I understand `dump-autoload` shouldn't make use of it.
I read through all the Composer docs but haven't found any option to investigate it deeper.
How can I pinpoint the origin of the problem?

**Comment by fredden** [comment]
What are the exact commands that are being run on production versus development? What version of Composer is in use in each environment?
I suspect that the production invocation is using `--optimize` whereas the development invocation isn't.

**Reply by fluffycondor** [reply]
Thank you for the answer.
Exact commands are `composer dump-autoload --profile -vvvvv` in both cases with no additional parameters.
Composer version in production and development is the same – `2.5.5`

**Reply by fredden** [reply]
Thanks for confirming. Are you able to share a `composer.json` file? It sounds like you've already checked for differences. I had a quick look through the code and can't see anything obvious at this stage. I'm assuming you're on the same version of PHP (8.0.15) in both environments, and that APCu is configured similarly, and that the environment variables (and contents of user-level Composer files) match where relevant.

**Reply by fluffycondor** [reply]
The PHP version is 8.0.15 on production and 8.0.13 on dev, I hope this doesn't make much difference.
Unfortanetely, production evnironment is out of my control, so I can't confirm they are matching :( Also not quite sure what does APCu means. Please let me know if possible what exactly could be checked there, so I could assign a task to our devops.
Here's `composer.json`:
```json
{
  "type": "project",
  "minimum-stability": "dev",
  "require": {
    "php": ">=8.0",
    "mpdf/mpdf": "8.0.*",
    "ext-bcmath": "*",
    "endroid/qr-code": "4.x-dev",
    "ext-iconv": "*",
    "ext-fileinfo": "*",
    "ext-pdo": "*"
  },
  "autoload": {
    "classmap": ["application/", "sparks/", "system/"]
  },
  "config": {
    "process-timeout": 1800
  },
  "require-dev": {
    "phpunit/phpunit": "^9",
    "vimeo/psalm": "5.x-dev",
    "phpstan/phpstan": "^1"
  }
}
```

**Comment by Seldaek** [comment, marked as answer]
Woah this is weird.. if you can't access it to debug where it spends all its time though it's gonna be really hard to help out.

I'm guessing either filesystem is horribly slow for some reason, or some of those 4100 files are gigantic and causing issues with the classmap scanner, but you should see that in dev too.

One thing that could change is that in prod you have some more files to scan like cache dirs or something within some of `"application/", "sparks/", "system/"` ?

**Reply by fluffycondor** [reply]
Thank you for the answer. Indeed, there's a cache folder inside `application/`. I've created a task to check what's inside there.
Any chances adding feature to output how many files were scanned during `dump-autoload` in verbose mode? It would be quite entertaining if this assumption actually turned out to be the root cause of the problem 😅

**Reply by Seldaek** [reply]
It will scan any `.php`, `.hh` or `.inc` file within those directories so you could hopefully run this before composer dump-autoload to see what the contents are:

`find application/ -type f \( -name "*.inc" -o -name "*.php" \) | wc -l`

But is the cache empty/somewhere else in development? Otherwise this wouldn't really explain it..

**Reply by fluffycondor** [reply]
Thank you for the help! I have pinpointed the issue.
There is indeed a cache folder inside `application/` folder. It has 42kk files 🙈, but none of them have `.php`, `.hh` or `.inc` extension though. Excluding this folder from the scanning helped out.
Looks like there's a bottleneck in the `symfony/finder` package, because `find` commands execute over `application/` folder relatively fast. It's because there's `\RecursiveDirectoryIterator` under the hood, that actually loops over every single file in the provided directory, and filters are applied only by decorating it with *FilterIterators, instead of just globbing files by a pattern or something. So the current algorithm is extemely inefficient for such cases. (I agree it's insane to have 42kk cache files in application folder though.)

**Reply by Seldaek** [reply]
Yeah that's a tricky one.. By 42kk do you mean 42 million files? 

https://getcomposer.org/doc/04-schema.md#exclude-files-from-classmaps wouldn't even help here as it filters out files *after* the symfony/finder is done finding files.

So I'd say the best solution is probably to use more fine-grained folders in your classmap, like application/src or whatever you can do to avoid including the cache folder. Or move the cache out of there.

**Reply by fluffycondor** [reply]
> By 42kk do you mean 42 million files?

Yes.
> the best solution is probably to use more fine-grained folders in your classmap, like application/src

Yeah, that's what I did. 
Thanks for helping to investigate the issue!
